# **I Solving the Sum of Two Digits Programming Challenge**

## C++

In [69]:
%%writefile AplusB.cpp
#include <iostream>

int sum_of_two_digits(int first_digit, int second_digit) {
    return first_digit + second_digit;
}

int main() {
    int a = 0;
    int b = 0;
    std::cin >> a;
    std::cin >> b;
    std::cout << sum_of_two_digits(a, b);
    return 0;
}

Writing AplusB.cpp


In [70]:
!g++ AplusB.cpp -o AplusB
!./AplusB

4 11
15

### Test input output files

In [72]:
%%writefile input1.txt
4 8

Overwriting input1.txt


In [73]:
%%writefile output1.txt
12

Writing output1.txt


In [74]:
%%writefile input2.txt
-5 9

Writing input2.txt


In [75]:
%%writefile output2.txt
4

Writing output2.txt


In [76]:
!./AplusB < input1.txt

12

In [77]:
!./AplusB < input2.txt

4

### Comparing files with grep

https://askubuntu.com/questions/546796/comparing-two-text-files

In [78]:
!grep -xvFf output1.txt output1.txt 

In [79]:
!grep -xvFf output1.txt output2.txt 

4


In [80]:
!grep -xvFf output2.txt output1.txt 

12


In [81]:
!./AplusB < input1.txt> ctest1.txt
!grep -xvFf output1.txt ctest1.txt

In [82]:
!./AplusB < input2.txt > ctest2.txt
!grep -xvFf output2.txt ctest2.txt

## Python

In [84]:
%%writefile sum.py
# python3

def sum_of_two_digits(first_digit, second_digit):
    return first_digit + second_digit

if __name__ == '__main__':
    a, b = map(int, input().split())
    print(sum_of_two_digits(a, b))

Writing sum.py


In [85]:
!python sum.py

32 90
122


In [86]:
!python AplusB.py  < input1.txt> ptest1.txt
!grep -xvFf output1.txt ptest1.txt

In [87]:
!python AplusB.py  < input2.txt> ptest2.txt
!grep -xvFf output2.txt ptest2.txt

# **II Solving The Maximum Pairwise Product Programming Challenge in C++**

## C++

In [64]:
%%writefile max_pairwise_product.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int MaxPairwiseProduct(const std::vector<int>& numbers) {
    int max_product = 0;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = first + 1; second < n; ++second) {
            max_product = std::max(max_product,
                numbers[first] * numbers[second]);
        }
    }

    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}


Writing max_pairwise_product.cpp


In [66]:
!g++ max_pairwise_product.cpp -o max_pairwise_product
!./max_pairwise_product

6
23 1 89 56 34 5 4984
4984


### Test 1: Small numbers

In [88]:
%%writefile input1.txt
10
1 2 3 4 5 6 7 8 9 10

Overwriting input1.txt


In [89]:
!./max_pairwise_product < input1.txt

90


##### Test 2: largest input numbers 200000 200000 in the input

In [90]:
%%writefile input2.txt
6
1 3 200000 9 200000 23

Overwriting input2.txt


In [91]:
!./max_pairwise_product < input2.txt

1345294336


Changing the variables to long long

In [92]:
%%writefile max_pairwise_product_long.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {
    long long max_product = 0LL;
	  long long aux_product = 0LL;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = first + 1; second < n; ++second) {
			       aux_product =  (long long) numbers[first] * numbers[second];
            if ( max_product < aux_product)  {
				        max_product = aux_product;
             }
        }
    }

    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}


Overwriting max_pairwise_product_long.cpp


In [93]:
!g++ max_pairwise_product_long.cpp -o max_pairwise_product_long
!./max_pairwise_product_long < input1.txt
!./max_pairwise_product_long < input2.txt

90
40000000000


### Test 3 stress tests 20 times random n in [2, 20]

In [94]:
%%writefile test3.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <ctime>
long long MaxPairwiseProduct(const std::vector<int>& numbers) {
    long long max_product = 0LL;
	  long long aux_product = 0LL;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = first + 1; second < n; ++second) {
			       aux_product =  (long long) numbers[first] * numbers[second];
            if ( max_product < aux_product)  {
				        max_product = aux_product;
             }
        }
    }

    return max_product;
}

int main()
{
    srand(time(NULL));
    int n, aleatorio;
    n = 20;
    //std::cout<<n<<"\n";
    for(int i=0; i < n; i++){
        int nums = rand()%(20-2 + 1) + 2;
        std::cout<<nums<<"\n";
        std::vector<int> numbers(nums);
        for (int j = 0; j < nums; j++)
        {
            numbers[j] = rand()%200000;
        }
        std::cout<<MaxPairwiseProduct(numbers)<<"\n";
    }
    return 0;
}

Overwriting test3.cpp


In [ ]:
!g++ test3.cpp -o test3

In [ ]:
!./test3 

16
35823345064
9
29677457844
10
33714683640
10
24357138968
13
30440346300
18
37009876995
6
17212293395
14
38749919375
14
35016580368
8
15854870140
7
39758510430
17
29175950800
17
35896802265
18
34281757530
4
21884714826
11
21192281850
8
15221009920
7
17679597952
12
28119510924
3
470147886


### Test 4 time test measure the execution time running with the the largest worst case file (200000 increasing numbers, 200000 and 200000 the end)

In [55]:
%%writefile Test2Write.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input3.txt");

 MyFile << 200000 <<"\n";  // n 
 for (int i=1; i < 200001; i++){
     MyFile << i <<" ";
 }
  
  // Close the file
  MyFile.close();
}

Overwriting Test2Write.cpp


In [95]:
!g++ Test2Write.cpp -o Test2Write
!./Test2Write

Measure time

https://askubuntu.com/questions/53444/how-can-i-measure-the-execution-time-of-a-terminal-process

In [ ]:
!g++ max_pairwise_product_long.cpp -o max_pairwise_product_long
!time ./max_pairwise_product_long < input3.txt

39999800000

real	2m9.177s
user	2m8.700s
sys	0m0.033s


####Calculate  the number of instructions executed by the algorithm as a function of n, add in each line a comment with the number of instructions that are executed in that line

```cpp
 int main() {                         //N inst exec by line executed  
   
    int n;                            //1                      
    std::cin >> n;                    //1
                                              
    std::vector<int> numbers(n);      //n    init vector
    .
    .
 }

```

In [ ]:
%%writefile test_init_vectors.cpp
#include <iostream>
#include <vector>
#include <algorithm>
int main() {                                                               
    int n = 10;                                                                                                                            //1
    std::vector<int> numbers(n);                                            
    for (int i = 0; i < n; ++i) {                                           
        std::cout << numbers[i];                                             
    }

    std::cout << "\n";                       
    return 0;                                                              
}  

Writing test_init_vectors.cpp


In [ ]:
!g++ test_init_vectors.cpp -o test_init_vectors
!./test_init_vectors

0000000000


In [ ]:
%%writefile max_pairwise_product_long.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {              //N inst exec by line executed 
    long long max_product = 0LL;                                             //1                                          
	  long long aux_product = 0LL;                                             //1 
    int n = numbers.size();                                                  //1 
                                                                             
    for (int first = 0; first < n; ++first) {                                //1+2n  
        for (int second = first + 1; second < n; ++second) {                 //n + (n-1)n 
			       aux_product =  (long long) numbers[first] * numbers[second];    //(n-1)n/2 
            if (max_product < aux_product)  {                                //(n-1)n/2 
				        max_product = aux_product;                                   //[0, (n-1)n/2] 
             }
        }
    }
                                                                             // -------------- 
    return max_product;                                                      // 4 + 3n + [2(n-1)n, 2.5 (n-1)n] 
}

int main() {                                              //N inst exec by line executed 
    int n;                                                //1  
    std::cin >> n;                                        //1
    std::vector<int> numbers(n);                          //n
    for (int i = 0; i < n; ++i) {                         //1 + 2n
        std::cin >> numbers[i];                           //n
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";     //1 + 1  + 4 + 3n + [2(n-1)n, 2.5 (n-1)n]
    return 0;                                             // -------------- 
}                                                         // 9 + 7n + [2(n-1)n, 2.5 (n-1)n]

Overwriting max_pairwise_product_long.cpp


Count executed intructions in code and verify with inputs 1, 2 and 3

In [ ]:
%%writefile max_pairwise_product_long_execints.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers, long long &execints) {              //N inst exec by line executed 
    
    long long m = 0LL;                                                       //1
    
    long long max_product = 0LL;                                             //1
    execints++;

	  long long aux_product = 0LL;                                             //1
    execints++;

    int n = numbers.size();                                                  //1 
    execints++;

    execints++;                                                                         
    for (int first = 0; first < n; ++first) {                                //1+2n  
        execints += 2;

        long long m = 0LL;

        execints++;                                    
        for (int second = first + 1; second < n; ++second) {                 //n+ (n-1)n 
            execints += 2;                                                

			      aux_product =  (long long) numbers[first] * numbers[second];    //(n-1)n/2 
            execints++; 

            execints++; 
            if ( max_product < aux_product)  {                               //(n-1)n/2 
                       
				        max_product = aux_product;                                   //[0, (n-1)n/2] 
                execints++; 
                m++;
             }
        }
        if ( (first <= 10) || (n- first <= 10)) std::cout << m << "\n";
    }
                                                                            // -------------- 
    return max_product;                                                     // 4 + 3n + [2(n-1)n, 2.5 (n-1)n] 
}

int main() {                                                 //N inst exec by line executed 
    long long execints =0LL;

    int n;                                                   //1
    execints++;                                                              
    
    std::cin >> n;                                           //1
    execints++;

    std::vector<int> numbers(n);                             //n
    execints += n;
    
    execints++;
    for (int i = 0; i < n; ++i) {                            //1 + 2n
        execints += 2;

        std::cin >> numbers[i];
        execints++;                                          //n
    }
    
    //std::cout << MaxPairwiseProduct(numbers) << "\n";      //1 + 1  + 4 + 3n + [2(n-1)n, 2.5 (n-1)n]

    execints++;
    int M = MaxPairwiseProduct(numbers,execints);

    execints++;
    std::cout <<  execints << "\n";                       
    return 0;                                                // -------------- 
}                                                            // 9 + 7n + [2(n-1)n, 2.5 (n-1)n]

Writing max_pairwise_product_long_execints.cpp


In [ ]:
!g++ max_pairwise_product_long_execints.cpp -o max_pairwise_product_long_execints
!./max_pairwise_product_long_execints < input1.txt
!./max_pairwise_product_long_execints < input2.txt
!./max_pairwise_product_long_execints < input4.txt

9
5
4
3
2
2
2
2
1
0
289
2
1
2
0
0
0
116
200000
100001
66667
50001
40001
33334
28572
25001
22223
20001
18182
2
2
2
2
2
2
2
2
1
0
80004472127


The number of instructions executed by the algooirithm with a list of size $n$ is  

$$T = 9 + 7n + [2(n-1)n, 2.5 (n-1)n]$$

It depends on the size of the lists and  the order of the list.

For the input1.txt 
```
10
1 2 3 4 5 6 7 8 9 10
```
$$T = 9 + 7(10) + 2(9)(10) +  9 + 5 + 4 + 3 + 2 + 2 + 2 + 2 + 1 + 0  = 289 $$

For the input2.txt 
```
6
1 3 200000 9 200000 23
```
$$T = 9 + 7(6) + 2(5)(6)  +  2 + 1 + 2 + 0 + 0 + 0 = 116 $$


For the input3.txt that is the worst case (n=200000 and list in incresing order)

```
200000
1 2 3 4 5 6 7 8 9 10 ... 199998 199999 200000
```

$$
\begin{array}{ll}
T &= 9 + 7(200000) + (2)(199999)(200000)  + (199999 +
100000 + 66667 +50000 \cdots 2 + 2 + 1)\\
&= 80003672077  \\
  &< 9 + 7(200000) + (2.5)(199999)(200000) \\
  & = 100000900009
\end{array}
  $$

Calculate estimated time by instruction and number of instructions by second

In [ ]:
# Estimated time by instruction

128/80003672077 


1.5999265618308827e-09

In [ ]:
# Estimated number of instructions by second

1/(128/80003672077) 

625028688.1015625

##### Nueva sección

### **Efficient solution**

Add in each line a coment with the number of intstrcution that are executed in that line

In [124]:
%%writefile max_pairwise_product_eficiente.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <list>

long long MaxPairwiseProductEficiente(const std::vector<int>& numbers, long long &execints) {
    int n = numbers.size();                                             //1
    execints++;

    long long max_product = 0LL;                                        //1
    execints++;

    long long mayor = 0;                                                //1
    execints++;

    long long smayor = 0;                                               //1
    execints++;

    execints++;
    long long co1 = 0LL;
    long long co2 = 0LL;
    long long co3 = 0LL;  
    for(int i = 0; i < n; i++){                                         //1+2n
        execints+=2;

        
        execints++;                                                     //n
        if(numbers[i]>mayor){
          
          smayor = mayor;                                             //[0,2n]
          execints++;

          mayor = numbers[i];                                         
          execints++;

          co1++;

        }else{                                                       //[0,n]
            co2++;
                                                                
            execints++;
            if(numbers[i]>smayor){   
                co3++;

                smayor = numbers[i];                                  //[0,n]
                execints++;
            }   
        }
        
    }
    //std::cout<<"co1: "<<co1<<"\n";
    //std::cout<<"co2: "<<co2<<"\n";
    //std::cout<<"co3: "<<co3<<"\n";
    max_product = (long long)mayor*smayor;                              //1
    execints++;
                                                              //--------------
    return max_product;                                       //  6 + [3n,8n]
}

int main() {         
    long long execints =0LL;                                              

    int n;                                                              //1
    execints++;

    std::cin >> n;                                                      //1
    execints++;                                                    

    std::vector<int> numbers(n);                                        //n
    execints+=n;   
                                        
    long long maxproduc = 0LL;                                          //1
    execints++;


    execints++;
    for (int i = 0; i < n; ++i) {                                       //1 + 2n
        execints+=2;

        std::cin >> numbers[i];                                         //n
        execints++;

    }

    maxproduc = MaxPairwiseProductEficiente(numbers,execints);          //  6 + [3n,8n]
    execints++;                                                         //1

    std::cout<<maxproduc<<"\n";       
    
    //std::cout <<"execints: "<<execints << "\n";                          //-----------------
    return 0;                                               //    11 + [7n, 12n]
}



Overwriting max_pairwise_product_eficiente.cpp


In [125]:
!g++ max_pairwise_product_eficiente.cpp -o max_pairwise_product_eficiente

Test with inputs 1 and 2

In [126]:
!./max_pairwise_product_eficiente < input1.txt
!./max_pairwise_product_eficiente < input2.txt

90
40000000000


Test and time it with input 3

In [164]:
!time ./max_pairwise_product_eficiente < input3.txt

39999800000

real	0m0.067s
user	0m0.058s
sys	0m0.004s


20 stress tests

In [179]:
%%writefile stresstest_efficient.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <ctime>
long long MaxPairwiseProductEficiente(const std::vector<int>& numbers) {
    int n = numbers.size();
    long long max_product = 0LL;
    long long mayor = 0;
    long long smayor = 0;
    for(int i = 0; i < n; i++){
        if(numbers[i]>mayor){
          smayor = mayor;
          mayor = numbers[i];
        }else{
          if(numbers[i]>smayor){
            smayor = numbers[i];
          }  
        } 
    }
    max_product = (long long)mayor*smayor;
    return max_product;
}

int main()
{
    srand(time(NULL));
    int n, aleatorio;
    n = 20;
    //std::cout<<n<<"\n";
    for(int i=0; i < n; i++){
        int nums = rand()%(20-2 + 1) + 2;
        std::cout<<nums<<"\n";
        std::vector<int> numbers(nums);
        for (int j = 0; j < nums; j++)
        {
            numbers[j] = rand()%200000;
        }
        std::cout<<MaxPairwiseProductEficiente(numbers)<<"\n";
        
    }
    return 0;
}

Overwriting stresstest_efficient.cpp


In [180]:
!g++ stresstest_efficient.cpp -o stresstest_efficient

In [181]:
!./stresstest_efficient

8
28418862031
15
31302976218
11
34346906240
17
36109426201
12
34319994683
10
21667398402
20
37718852544
13
33313382173
13
35644342650
4
18853581824
6
24124854017
5
26325662102
20
38999465593
19
36455695496
7
11454562684
2
34763342908
17
36699461469
20
36308514060
9
30175159304
9
30854445111


Count executed instructions in code and verify with inputs 1,2 and 3

In [131]:
%%writefile max_pairwise_product_eficiente_execints.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <list>

long long MaxPairwiseProductEficiente(const std::vector<int>& numbers, long long &execints) {
    int n = numbers.size();                                             //1
    execints++;

    long long max_product = 0LL;                                        //1
    execints++;

    long long mayor = 0;                                                //1
    execints++;

    long long smayor = 0;                                               //1
    execints++;

    execints++;
    long long co1 = 0LL;
    long long co2 = 0LL;
    long long co3 = 0LL;  
    for(int i = 0; i < n; i++){                                         //1+2n
        execints+=2;

        
        execints++;                                                     //n
        if(numbers[i]>mayor){
          
          smayor = mayor;                                             //[0,2n]
          execints++;

          mayor = numbers[i];                                         
          execints++;

          co1++;

        }else{                                                       //[0,n]
            co2++;
                                                                
            execints++;
            if(numbers[i]>smayor){   
                co3++;

                smayor = numbers[i];                                  //[0,n]
                execints++;
            }   
        }
        
    }
    std::cout<<"co1: "<<co1<<"\n";
    std::cout<<"co2: "<<co2<<"\n";
    std::cout<<"co3: "<<co3<<"\n";
    max_product = (long long)mayor*smayor;                              //1
    execints++;
                                                              //--------------
    return max_product;                                       //  6 + [3n,8n]
}

int main() {         
    long long execints =0LL;                                              

    int n;                                                              //1
    execints++;

    std::cin >> n;                                                      //1
    execints++;                                                    

    std::vector<int> numbers(n);                                        //n
    execints+=n;   
                                        
    long long maxproduc = 0LL;                                          //1
    execints++;


    execints++;
    for (int i = 0; i < n; ++i) {                                       //1 + 2n
        execints+=2;

        std::cin >> numbers[i];                                         //n
        execints++;

    }

    maxproduc = MaxPairwiseProductEficiente(numbers,execints);          //  6 + [3n,8n]
    execints++;                                                         //1

    //std::cout<<maxproduc<<"\n";       
    
    std::cout <<"execints: "<<execints << "\n";                          //-----------------
    return 0;                                               //    11 + [7n, 12n]
}



Writing max_pairwise_product_eficiente_execints.cpp


In [132]:
!g++ max_pairwise_product_eficiente_execints.cpp -o max_pairwise_product_eficiente_execints

In [133]:
!./max_pairwise_product_eficiente_execints<input1.txt

co1: 10
co2: 0
co3: 0
execints: 101


In [134]:
!./max_pairwise_product_eficiente_execints<input2.txt

co1: 3
co2: 3
co3: 2
execints: 64


In [135]:
!./max_pairwise_product_eficiente_execints<input3.txt

co1: 200000
co2: 0
co3: 0
execints: 1800011



For the input1.txt 
```
10
1 2 3 4 5 6 7 8 9 10
```
$$T = 11 + (7)(10) + (2)(10) +  (1)(0) + (1)(0) = 101 $$

For the input2.txt 
```
6
1 3 200000 9 200000 23
```
$$T = 11 + 7(6) + 2(3) + (1)(3) + (1)(2)= 64$$
 

For the input3.txt that is the worst case (n=200000 and list in incresing order)

```
200000
1 2 3 4 5 6 7 8 9 10 ... 199998 199999 200000
```

$$T = 11 + (7)(200000) + (2)(200000) + (1)(0) + (1)(0)= 1800011$$

Calculate estimated time by instruction and number of instructions by second

In [167]:
# Estimated time by instruction using input 3

0.067/1800011

3.7221994754476504e-08

In [168]:
# Estimated number of instructions by second using input 3

1/(0.067/1800011)

26865835.82089552

### Super efficient solution

Whitout storing the list in memory and whitout calling a function, add in each line a coment with the number of intstrcution that are executed in that line 

In [193]:
%%writefile max_pairwise_product_super_eficiente.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <list>


int main() {
    long long execints =0LL; 
    long long co1 = 0LL;
    long long co2 = 0LL;
    long long co3 = 0LL;  

    int n;                                                                //1
    execints++;

    long long max_product = 0LL;                                          //1
    execints++;

    long long mayor = 0LL;                                                //1
    execints++;

    long long smayor = 0LL;                                               //1
    execints++;

    long long temp = 0LL;                                                 //1
    execints++;

    std::cin >> n;                                                        //1
    execints++;

    execints++;
    for (int i = 0; i < n; ++i) {                                     //2n + 1
        execints+=2;

        std::cin >> temp;                                                 //1
        execints++;

        execints++;                                                       //1
        if(temp>mayor){                                               //[0,2n]
          smayor = mayor;
          execints++; 

          mayor = temp;
          execints++; 
          co1++;
        }else{                                                        //[0,n]
          co2++;
          execints++;                                                     
          if(temp>smayor){                                            //[0,n]
            co3++;
            smayor = temp;
            execints++;   

          }  
        } 
    }
    max_product = (long long)mayor*smayor;                            //1
    execints++;                                                     

    //std::cout <<"execints: "<<execints << "\n";        //-----------                              
    std::cout<<max_product<<"\n";
    return 0;                                                     // 10 + [2n, 6n]
}


Overwriting max_pairwise_product_super_eficiente.cpp


In [194]:
!g++ max_pairwise_product_super_eficiente.cpp -o max_pairwise_product_super_eficiente

Test with inputs 1 and 2

In [195]:
!./max_pairwise_product_super_eficiente<input1.txt
!./max_pairwise_product_super_eficiente<input2.txt

90
40000000000


Test and time it with input 3

In [215]:
!time ./max_pairwise_product_super_eficiente<input3.txt

39999800000

real	0m0.059s
user	0m0.056s
sys	0m0.003s


20 stress tests

In [185]:
%%writefile stresstest_superefficient.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <ctime>

int main()
{
    srand(time(NULL));
    int n, aleatorio;
    n = 20;
    //std::cout<<n<<"\n";
    for(int i=0; i < n; i++){
        long long max_product = 0LL;
        long long mayor = 0LL;
        long long smayor = 0LL;
        long long temp = 0LL;
        int nums = rand()%(20-2 + 1) + 2;
        std::cout<<nums<<"\n";
        for (int j = 0; j < nums; j++)
        {
            temp = rand()%200000;
            if(temp>mayor){
              smayor = mayor;
              mayor = temp;
            }else if(temp>smayor){
              smayor = temp;
            }

        }
        max_product = (long long)mayor*smayor;
        std::cout<<max_product<<"\n";
        
    }
    return 0;
}

Overwriting stresstest_superefficient.cpp


In [186]:
!g++ stresstest_superefficient.cpp -o stresstest_superefficient

In [187]:
!./stresstest_superefficient

14
36729293475
4
24540192936
9
30025920265
13
38155486278
19
35916486106
5
19494903714
8
29224940473
18
36066361305
11
22497822304
9
18303249875
11
23159973580
10
32317159770
10
27013941500
4
9120459960
14
36551117228
6
19847563047
7
31331756103
17
35277837777
17
25547837669
2
7845813335


Count executed instructions in code and verify with inputs 1,2 and 3

In [208]:
%%writefile max_pairwise_product_super_eficiente_execints.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <list>


int main() {
    long long execints =0LL; 
    long long co1 = 0LL;
    long long co2 = 0LL;
    long long co3 = 0LL;  

    int n;                                                                //1
    execints++;

    long long max_product = 0LL;                                          //1
    execints++;

    long long mayor = 0LL;                                                //1
    execints++;

    long long smayor = 0LL;                                               //1
    execints++;

    long long temp = 0LL;                                                 //1
    execints++;

    std::cin >> n;                                                        //1
    execints++;

    execints++;
    for (int i = 0; i < n; ++i) {                                     //2n + 1
        execints+=2;

        std::cin >> temp;                                                 //n
        execints++;

        execints++;                                                       //n
        if(temp>mayor){                                               //[0,2n]
          smayor = mayor;
          execints++; 

          mayor = temp;
          execints++; 
          co1++;
        }else{                                                        //[0,n]
          co2++;
          execints++;                                                     
          if(temp>smayor){                                            //[0,n]
            co3++;
            smayor = temp;
            execints++;   

          }  
        } 
    }

    max_product = (long long)mayor*smayor;                            //1
    execints++;                                                     

    std::cout<<"co1: "<<co1<<"\n";
    std::cout<<"co2: "<<co2<<"\n";
    std::cout<<"co3: "<<co3<<"\n";

    std::cout <<"execints: "<<execints << "\n";        //-----------                              
    //std::cout<<max_product<<"\n";
    return 0;                                                     // 8 + [4n, 8n]
}


Overwriting max_pairwise_product_super_eficiente_execints.cpp


In [209]:
!g++ max_pairwise_product_super_eficiente_execints.cpp -o max_pairwise_product_super_eficiente_execints

In [213]:
! ./max_pairwise_product_super_eficiente_execints<input1.txt

co1: 10
co2: 0
co3: 0
execints: 68


In [211]:
! ./max_pairwise_product_super_eficiente_execints<input2.txt

co1: 3
co2: 3
co3: 2
execints: 43


In [212]:
! ./max_pairwise_product_super_eficiente_execints<input3.txt

co1: 200000
co2: 0
co3: 0
execints: 1200008



For the input1.txt 
```
10
1 2 3 4 5 6 7 8 9 10
```
$$T = 8 + (4)(10) + (2)(10) +  (1)(0) + (1)(0) = 68 $$

For the input2.txt 
```
6
1 3 200000 9 200000 23
```
$$T = 8 + (4)(6) + 2(3) + (1)(3) + (1)(2)= 43$$
 

For the input3.txt that is the worst case (n=200000 and list in incresing order)

```
200000
1 2 3 4 5 6 7 8 9 10 ... 199998 199999 200000
```

$$T = 8 + (4)(200000) + (2)(200000) + (1)(0) + (1)(0)= 1200008$$

Calculate estimated time by instruction and number of instructions by second

In [216]:
# Estimated time by instruction using input 3

0.059/1200008

4.9166338891074055e-08

In [217]:
# Estimated number of instructions by second using input 3

1/(0.059/1200008)

20339118.644067798

## Python

In [234]:
%%writefile max_pairwise_product1.py
# python3

def max_pairwise_product(numbers):
    n = len(numbers)
    max_product = 0
    for first in range(n):
        for second in range(first + 1, n):
            max_product = max(max_product,
                numbers[first] * numbers[second])

    return max_product


if __name__ == '__main__':
    input_n = int(input())
    input_numbers = [int(x) for x in input().split()]
    print(max_pairwise_product(input_numbers))

Overwriting max_pairwise_product1.py


In [235]:
!python max_pairwise_product1.py

6
9 8 7 6 5 4 
72


### Test 1: Small numbers

In [225]:
!python max_pairwise_product1.py < input1.txt

90


##### Test 2: largest input numbers 200000 200000 in the input

In [226]:
!python max_pairwise_product1.py< input2.txt

40000000000


### Test 3 stress tests 20 times random n in [2, 20]

In [229]:
%%writefile max_pairwise_product_stressTest.py
# python3
import numpy as np
def max_pairwise_product(numbers):
    n = len(numbers)
    max_product = 0
    for first in range(n):
        for second in range(first + 1, n):
            max_product = max(max_product,
                numbers[first] * numbers[second])

    return max_product

if __name__ == '__main__':
  n = 20
  for i in range(n):
    nums = np.random.randint(18, size = 1)+2
    print(nums[0],"\n")
    numbers = np.random.randint(199999, size=nums)
    numbers = numbers + 1
    print( max_pairwise_product(numbers),"\n")


Overwriting max_pairwise_product_stressTest.py


In [230]:
!python max_pairwise_product_stressTest.py

16 

29471805160 

7 

27106707085 

14 

29777225265 

14 

27801347445 

18 

26599747068 

9 

26925324580 

4 

16453734479 

15 

37256600476 

11 

27744770460 

9 

16747544037 

15 

28974301845 

3 

31114582317 

15 

23789096045 

17 

35551642866 

15 

28206806808 

2 

10959755409 

8 

36159959250 

9 

27588307097 

12 

36302383744 

19 

34688419601 



### Test 4 time test measure the execution time running with the the largest worst case file (200000 increasing numbers, 200000 and 200000 the end)

In [ ]:
!time python max_pairwise_product1.py<input3.txt

40000000000

real	84m44.197s
user	84m3.897s
sys	0m1.346s


### **Efficient solution**

Add in each line a coment with the number of intstrcution that are executed in that line

In [326]:
%%writefile max_pairwise_product1_eficiente.py
# python3


def max_pairwise_product_eficiente(numbers):
    co1 = 0
    co2 = 0
    co3 = 0
    global execints
    n = len(numbers)                                                #1
    execints+=1

    max_product = 0                                                 #1    
    execints+=1

    mayor = 0                                                       #1
    execints+=1

    smayor=0                                                        #1
    execints+=1

    for first in range(n):                                          #n
      execints+=1

      execints+=1                                                   #n
      if numbers[first]>mayor:                                     #[0,2n]
        smayor = mayor
        execints+=1

        mayor = numbers[first]
        execints+=1
        co1+=1
      else:                                                        #[0,n]

        co2+=1
        execints+=1                                                
        if numbers[first]>smayor:                                  #[0,n]
          co3+=1
          smayor = numbers[first]
          execints+=1  

      #print("Mayor: ",mayor,"  -  Segundo: ",smayor,"\n")
    max_product = mayor*smayor                                      #1
    execints +=1
    
    #print("co1: ",co1,"\n")
    #print("co2: ",co2,"\n")
    #print("co3: ",co3,"\n") 
    
    
                                                            #---------------
    return max_product                                      #5+[2n,6n]


if __name__ == '__main__':
    execints = int()
    input_n = int(input())                                    #1
    execints = execints +1

    input_numbers = [int(x) for x in input().split()]         #n
    execints += input_n
                                                      
    execints = execints +1                                      #1
                                                                  #-----------
    max_prod = (max_pairwise_product_eficiente(input_numbers))      #7+[2n,6n]
    print(max_prod)
    #print("execints: ",execints,"\n")                        #------------
                                                             
                                                              #7+[3n,7n]


Overwriting max_pairwise_product1_eficiente.py


Test with inputs 1 and 2

In [327]:
!python max_pairwise_product1_eficiente.py<input1.txt
!python max_pairwise_product1_eficiente.py<input2.txt

90
40000000000


Test and time it with input 3

In [328]:
! time python max_pairwise_product1_eficiente.py<input3.txt

39999800000

real	0m0.250s
user	0m0.204s
sys	0m0.044s


20 stress tests

In [284]:
%%writefile efficient_stressTest.py
# python3
import numpy as np
def max_pairwise_product_eficiente(numbers):
    n = len(numbers)
    max_product = 0
    mayor = 0
    smayor=0
    for first in range(n):
      if numbers[first]>mayor:
        smayor = mayor
        mayor = numbers[first]
      else:
        if numbers[first]>smayor:
          smayor = numbers[first]
      #print("Mayor: ",mayor,"  -  Segundo: ",smayor,"\n")
    max_product = mayor*smayor
    return max_product

if __name__ == '__main__':
  n = 20
  for i in range(n):
    nums = np.random.randint(18, size = 1)+2
    print(nums[0],"\n")
    numbers = np.random.randint(199999, size=nums)
    numbers = numbers + 1
    print( max_pairwise_product_eficiente(numbers),"\n")
    

Overwriting efficient_stressTest.py


In [285]:
!python efficient_stressTest.py

10 

29638140476 

16 

37301086318 

12 

35494517022 

9 

38244078990 

4 

12502544600 

19 

34480075368 

13 

39297011875 

14 

21634705388 

4 

27566871384 

6 

32036090073 

12 

23259045996 

15 

31265196170 

11 

33773164332 

14 

35696625349 

15 

32091252675 

8 

27113136040 

9 

24888050420 

14 

35969619658 

5 

16396322985 

8 

33631627680 



Count executed instructions in code and verify with inputs 1,2 and 3

In [325]:
%%writefile max_pairwise_product1_eficiente_excints.py
# python3


def max_pairwise_product_eficiente(numbers):
    co1 = 0
    co2 = 0
    co3 = 0
    global execints
    n = len(numbers)                                                #1
    execints+=1

    max_product = 0                                                 #1    
    execints+=1

    mayor = 0                                                       #1
    execints+=1

    smayor=0                                                        #1
    execints+=1

    for first in range(n):                                          #n
      execints+=1

      execints+=1                                                   #n
      if numbers[first]>mayor:                                     #[0,2n]
        smayor = mayor
        execints+=1

        mayor = numbers[first]
        execints+=1
        co1+=1
      else:                                                        #[0,n]

        co2+=1
        execints+=1                                                
        if numbers[first]>smayor:                                  #[0,n]
          co3+=1
          smayor = numbers[first]
          execints+=1  

      #print("Mayor: ",mayor,"  -  Segundo: ",smayor,"\n")
    
    max_product = mayor*smayor                                      #1
    execints+=1
    
    print("co1: ",co1,"\n")
    print("co2: ",co2,"\n")
    print("co3: ",co3,"\n") 
    
    
                                                            #---------------
    return max_product                                      #5+[2n,6n]


if __name__ == '__main__':
    execints = int()
    input_n = int(input())                                    #1
    execints = execints +1

    input_numbers = [int(x) for x in input().split()]         #n
    execints += input_n
                                                      
    execints = execints +1                                      #1
                                                                  #-----------
    max_prod = (max_pairwise_product_eficiente(input_numbers))      #7+[2n,6n]
    #print(max_prod)
    print("execints: ",execints,"\n")                        #------------
                                                             
                                                              #7+[3n,7n]

Overwriting max_pairwise_product1_eficiente_excints.py


In [329]:
!python max_pairwise_product1_eficiente_excints.py<input1.txt

co1:  10 

co2:  0 

co3:  0 

execints:  57 



In [330]:
!python max_pairwise_product1_eficiente_excints.py<input2.txt

co1:  3 

co2:  3 

co3:  2 

execints:  36 



In [331]:
!python max_pairwise_product1_eficiente_excints.py<input3.txt

co1:  200000 

co2:  0 

co3:  0 

execints:  1000007 




For the input1.txt 
```
10
1 2 3 4 5 6 7 8 9 10
```
$$T = 7 + (3)(10) + (2)(10) +  (1)(0) + (1)(0) = 57 $$

For the input2.txt 
```
6
1 3 200000 9 200000 23
```
$$T = 7 + 3(6) + 2(3) + (1)(3) + (1)(2)= 36$$
 

For the input3.txt that is the worst case (n=200000 and list in incresing order)

```
200000
1 2 3 4 5 6 7 8 9 10 ... 199998 199999 200000
```

$$T = 7 + (3)(200000) + (2)(200000) + (1)(0) + (1)(0)= 1000007 $$

Calculate estimated time by instruction and number of instructions by second

In [332]:
# Estimated time by instruction using input 3

0.250/1000007

2.499982500122499e-07

In [333]:
# Estimated number of instructions by second using input 3

1/(0.250/1000007)

4000028.0000000005

### Super efficient solution

Whitout storing the list in memory and whitout calling a function, add in each line a coment with the number of intstrcution that are executed in that line 

In [354]:
%%writefile max_pairwise_product1_super_efficient.py
# python3

if __name__ == '__main__':
    
    execints = int()
    co1 = int()
    co2 = int()
    co3 = int()
    
    max_product = 0                       #1
    execints+=1

    mayor = 0                             #1
    execints+=1

    smayor = 0                            #1
    execints+=1

    temp = 0                              #1
    execints+=1

    input_n = int(input())                #1
    execints+=1

    for x in input().split():             #n
      execints+=1

      temp = int(x)                       #n
      execints+=1

      execints+=1                         #n
      if temp>mayor:                    #[0, 2n]
        smayor = mayor
        execints+=1

        mayor = temp
        execints+=1
        co1+=1
      else:                              #[0,n]
        co2+=1
        execints+=1
        if temp>smayor:                  #[0,n]
          co3+=1
          execints+=1
          smayor = temp
    
    max_product = mayor*smayor           #1
    execints+=1

                                    #-----------------
                                    # 6+[3n,7n]

    #print("co1: ",co1,"\n")
    #print("co2: ",co2,"\n")
    #print("co3: ",co3,"\n") 
    #print("execints: ",execints,"\n") 
    print(max_product)



Overwriting max_pairwise_product1_super_efficient.py


Test with inputs 1 and 2

In [355]:
!python max_pairwise_product1_super_efficient.py<input1.txt
!python max_pairwise_product1_super_efficient.py<input2.txt

90
40000000000


Test and time it with input 3

In [376]:
!time python max_pairwise_product1_super_efficient.py<input3.txt

39999800000

real	0m0.239s
user	0m0.206s
sys	0m0.034s


20 stress tests

In [341]:
%%writefile super_efficient_stressTest.py
# python3
import numpy as np

if __name__ == '__main__':
  n = 20
  for i in range(n):
    nums = np.random.randint(18, size = 1)+2
    print(nums[0],"\n")
    numbers = np.random.randint(199999, size=nums)
    numbers = numbers + 1


    max_product = 0
    mayor = 0
    smayor = 0
    temp = 0
    for x in numbers:
      temp = int(x)
      if temp>mayor:
        smayor = mayor
        mayor = temp
      elif temp>smayor:
        smayor = temp
    max_product = mayor*smayor
    print(max_product)

Overwriting super_efficient_stressTest.py


In [342]:
!python super_efficient_stressTest.py

4 

23259759453
18 

31729518029
10 

28779218060
10 

22198161566
9 

20848446122
15 

34437339794
15 

33789920300
5 

15840695010
8 

34959333009
12 

36639940950
12 

29508277984
16 

29255321880
3 

23274738162
12 

17279965872
10 

24169800760
5 

21601636992
10 

25862992200
4 

17601619711
13 

25530138360
16 

38936936493


Count executed instructions in code and verify with inputs 1,2 and 3

In [350]:
%%writefile max_pairwise_product1_super_efficient_execints.py
# python3

if __name__ == '__main__':
    
    execints = int()
    co1 = int()
    co2 = int()
    co3 = int()
    
    max_product = 0                       #1
    execints+=1

    mayor = 0                             #1
    execints+=1

    smayor = 0                            #1
    execints+=1

    temp = 0                              #1
    execints+=1

    input_n = int(input())                #1
    execints+=1

    for x in input().split():             #n
      execints+=1

      temp = int(x)                       #n
      execints+=1

      execints+=1                         #n
      if temp>mayor:                    #[0, 2n]
        smayor = mayor
        execints+=1

        mayor = temp
        execints+=1
        co1+=1
      else:                              #[0,n]
        co2+=1
        execints+=1
        if temp>smayor:                  #[0,n]
          co3+=1
          execints+=1
          smayor = temp
    
    max_product = mayor*smayor           #1
    execints+=1

                                    #-----------------
                                    # 6+[3n,7n]

    print("co1: ",co1,"\n")
    print("co2: ",co2,"\n")
    print("co3: ",co3,"\n") 
    print("execints: ",execints,"\n") 
    #print(max_product)


Overwriting max_pairwise_product1_super_efficient_execints.py


In [357]:
!python max_pairwise_product1_super_efficient_execints.py<input1.txt

co1:  10 

co2:  0 

co3:  0 

execints:  56 



In [358]:
!python max_pairwise_product1_super_efficient_execints.py<input2.txt

co1:  3 

co2:  3 

co3:  2 

execints:  35 



In [359]:
!python max_pairwise_product1_super_efficient_execints.py<input3.txt

co1:  200000 

co2:  0 

co3:  0 

execints:  1000006 




For the input1.txt 
```
10
1 2 3 4 5 6 7 8 9 10
```
$$T = 6 + (3)(10) + (2)(10) +  (1)(0) + (1)(0) = 56 $$

For the input2.txt 
```
6
1 3 200000 9 200000 23
```
$$T = 6 + (3)(6) + 2(3) + (1)(3) + (1)(2)= 35$$
 

For the input3.txt that is the worst case (n=200000 and list in incresing order)

```
200000
1 2 3 4 5 6 7 8 9 10 ... 199998 199999 200000
```

$$T = 6 + (3)(200000) + (2)(200000) + (1)(0) + (1)(0)= 1000006$$

Calculate estimated time by instruction and number of instructions by second

In [378]:
# Estimated time by instruction using input 3

0.239/1000006

2.3899856600860394e-07

In [379]:
# Estimated number of instructions by second using input 3

1/(0.239/1000006)

4184125.5230125524

# **Screenhsoots of the Coursera Submissions in C++ and Python**

<img src ="https://raw.githubusercontent.com/juarodriguezc/AlgorithmsUN2021I/master/Lab6/juarodriguezc.JPG">

<img src = "https://raw.githubusercontent.com/juarodriguezc/AlgorithmsUN2021I/master/Lab6/juarodriguezc2.JPG">